### PR TITLE
EB push to cpp

### PR DESCRIPTION
**Pull request by dpgrote**

### CODE REVIEW

File: Source/FieldSolver/WarpX_FDTD.H [hunk 1]
```diff
@@ -5,7 +5,376 @@
...
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+void warpx_push_bz_yee(int i, int j, int k, Array4<Real> const& Bz,
+                       Array4<Real const> const& Ex, Array4<Real const> const& Ey,
+                       Real dtsdx, Real dtsdy, Real dxinv, Real rmin)
+{
+#if (AMREX_SPACEDIM == 3)
+    Bz(i,j,k) += - dtsdx * (Ey(i+1,j  ,k) - Ey(i,j,k))
+                 + dtsdy * (Ex(i  ,j+1,k) - Ex(i,j,k));
+#else
+#ifndef WARPX_RZ
```
Comment (MaxThevenet) [inline Source/FieldSolver/WarpX_FDTD.H]: I think we are trying to move away from nest preprocessor directives (in this case, `#ifndef WARPX_RZ` is inside the test on `(AMREX_SPACEDIM == 3)`).

@WeiqunZhang introduced variables `WARPX_DIM_RZ` to this effect (see commit a5df2874d). Here I think you could replace this piece of code with
```
#if   defined WARPX_DIM_3D
    Bz(i,j,k) += - dtsdx * (Ey(i+1,j  ,k) - Ey(i,j,k))
                 + dtsdy * (Ex(i  ,j+1,k) - Ex(i,j,k));
#elif defined WARPX_DIM_2D
    Bz(i,j,0) += - dtsdx * (Ey(i+1,j,0) - Ey(i,j,0));
#elif defined WARPX_DIM_RZ
    Real ru = 1. + 0.5/(rmin*dxinv + i + 0.5);
    Real rd = 1. - 0.5/(rmin*dxinv + i + 0.5);
    Bz(i,j,0) += - dtsdx * (ru*Ey(i+1,j,0) - rd*Ey(i,j,0));
#endif
```
Do you agree with this change? If so, could you include it in the rest of this PR?
@RemiLehe should we make this standard in WarpX?
Thanks!

File: Source/FortranInterface/WarpX_picsar.F90 [hunk 2]
```diff
@@ -2,33 +2,20 @@
...
 #define WRPX_PXR_RZ_VOLUME_SCALING_J      apply_rz_volume_scaling_j
-#define WRPX_PXR_PUSH_BVEC                pxrpush_emrz_bvec
-#define WRPX_PXR_PUSH_EVEC                pxrpush_emrz_evec
 
 #else
 
 #define WRPX_PXR_GETEB_ENERGY_CONSERVING  geteb2dxz_energy_conserving_generic
 #define WRPX_PXR_CURRENT_DEPOSITION       depose_jxjyjz_generic_2d
-#define WRPX_PXR_PUSH_BVEC                pxrpush_em2d_bvec
-#define WRPX_PXR_PUSH_EVEC                pxrpush_em2d_evec
```
Comment (MaxThevenet) [inline Source/FortranInterface/WarpX_picsar.F90]: Great to get rid of so many of these!

File: Source/FieldSolver/WarpX_FDTD.H (hunk 1 above)
Comment (dpgrote) [reply Source/FieldSolver/WarpX_FDTD.H]: The changes have been implemented.

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 3]
```diff
@@ -41,6 +42,7 @@ WarpX::EvolveB (int lev, PatchType patch_type, amrex::Real a_dt)
     const int patch_level = (patch_type == PatchType::fine) ? lev : lev-1;
     const std::array<Real,3>& dx = WarpX::CellSize(patch_level);
     Real dtsdx = a_dt/dx[0], dtsdy = a_dt/dx[1], dtsdz = a_dt/dx[2];
+    Real dxinv = 1./dx[0];
```
Comment (ax3l) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]: these constants should be constant :)
```suggestion
    const Real dxinv = 1./dx[0];
```

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 4]
```diff
@@ -192,6 +227,7 @@ WarpX::EvolveE (int lev, PatchType patch_type, amrex::Real a_dt)
     int patch_level = (patch_type == PatchType::fine) ? lev : lev-1;
     const std::array<Real,3>& dx = WarpX::CellSize(patch_level);
     Real dtsdx_c2 = c2dt/dx[0], dtsdy_c2 = c2dt/dx[1], dtsdz_c2 = c2dt/dx[2];
+    Real dxinv = 1./dx[0];
```
Comment (ax3l) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]:
```suggestion
    const Real dxinv = 1./dx[0];
```

File: Source/FieldSolver/WarpX_FDTD.H [hunk 5]
```diff
@@ -5,50 +5,411 @@
...
+                       Array4<Real const> const& Ex, Array4<Real const> const& Ey,
+                       Real dtsdx, Real dtsdy, Real dxinv, Real rmin)
+{
+#if defined WARPX_DIM_3D
+    Bz(i,j,k) += - dtsdx * (Ey(i+1,j  ,k) - Ey(i,j,k))
+                 + dtsdy * (Ex(i  ,j+1,k) - Ex(i,j,k));
+#elif defined WARPX_DIM_2D
+    Bz(i,j,0) += - dtsdx * (Ey(i+1,j,0) - Ey(i,j,0));
+#elif defined WARPX_DIM_RZ
+    Real ru = 1. + 0.5/(rmin*dxinv + i + 0.5);
```
Comment (ax3l) [inline Source/FieldSolver/WarpX_FDTD.H]: `const` you, `const` me, `const` us together... 🎶

File: Source/FieldSolver/WarpX_FDTD.H [hunk 6]
```diff
@@ -5,50 +5,411 @@
...
+
+static void warpx_calculate_ckc_coefficients(Real dtsdx, Real dtsdy, Real dtsdz,
+                                             Real &betaxy, Real &betaxz, Real &betayx, Real &betayz, Real &betazx, Real &betazy,
+                                             Real &gammax, Real &gammay, Real &gammaz,
+                                             Real &alphax, Real &alphay, Real &alphaz)
+{
+    // Cole-Karkkainen-Cowan push
+    // computes coefficients according to Cowan - PRST-AB 16, 041303 (2013)
+#if defined WARPX_DIM_3D
+    Real delta = std::max( { dtsdx,dtsdy,dtsdz } );
```
Comment (ax3l) [inline Source/FieldSolver/WarpX_FDTD.H]: `r...` can be `const` :)

**Pull request merged**